srcts/src/components/webcomponents/layoutColumns.ts

```diff
@@ -0,0 +1,560 @@
+ * 12-column grid. E.g. for 7 items we can use 14 units per row, with 7 column
+ * units per item on small screens and 2 column units per item on large screens.
+ * This choice is currently invariant to `nItems`.
+ *
+ * @param nItems The number of children in the layout.
+ * @param preferWider Whether to prefer wider columns (for smaller breakpoints).
+ * @param units The number of column units in a row.
+ * @returns The best-fit column widths.
+ */
+function bestFitColumnWidths(
+  nItems: number,
+  preferWider = false,
+  units: number | [null] | null = null
+): BestFitColumnWidths {
+  const fit = { units: units, widths: [0] } as BestFitColumnWidths;
+
+  if (isNA(fit.units)) {
+    // We're auto-fitting at all breakpoints, we can decide column units/row
+    fit.units = nItems > 7 ? 12 : nItems > 3 ? nItems * 2 : nItems;
+
+    if (nItems < 4) {
+      fit.widths = [1];
+      return fit;
+    }
+
+    if (nItems <= 7) {
+      fit.widths = [preferWider ? nItems : 2];
+      return fit;
+    }
+  }
+
+  // In fixed 12-column mode, we manually pick columns widths for small nItems
+  // that look better than the default auto-fit widths.
+  if (fit.units === 12) {
+    if (nItems <= 3) {
+      fit.widths = [[12, 6, 4][nItems - 1]];
+      return fit;
+    }
+
+    // nItems === 4, default auto-fit is fine (6 units wide, 3 units narrow)
+
+    if (nItems === 5 || nItems === 7) {
+      fit.widths = [preferWider ? 4 : 3];
+      return fit;
+    }
+
+    if (nItems === 6) {
+      fit.widths = [preferWider ? 4 : 2];
+      return fit;
+    }
+  }
+
+  const fctrs = preferWider ? [6, 4, 3] : [2, 3, 4];
+
+  const unitsItems = fctrs.map((x) => x * nItems);
+  const rows = unitsItems.map((val) => Math.ceil(val / 12));
+  const unitsTotal = rows.map((val) => val * 12);
+  const unitsEmpty = unitsTotal.map((val, idx) => val - unitsItems[idx]);
+
+  fit.widths = [fctrs[unitsEmpty.indexOf(Math.min(...unitsEmpty))]];
+
+  return fit;
+}
+
+/**
+ * Given a resolved BreakpointMap (i.e. auto-fit breakpoints have been
+ * determined), create a BreakpointColumnSpec.
+ *
+ * @param breaks The resolved BreakpointMap.
+ * @returns The BreakpointColumnSpec.
+ */
+function newBreakpointColumnSpec(
+  breaks: BreakpointMapResolved
+): BreakpointColumnSpec {
+  // throw if breaks isn't a breakpoint map
+  if (!(breaks instanceof Map)) {
+    throw new Error("Column widths must be specified as a Map or an object.");
+  }
+
+  const spec = new Map() as BreakpointColumnSpec;
+
+  for (const [breakName, bk] of breaks) {
+    if (bk.some((val) => val === 0)) {
+      throw new Error(
+        "Column values must be greater than 0 to indicate width, or negative to indicate a column offset."
+      );
+    }
+
+    if (bk.length > 1 && bk.some((val) => isNaN(val))) {
+      throw new Error(
+        "Cannot mix widths and missing values. All column widths must be specified, or choose auto widths using a single `null` value."
+      );
+    }
```

Comment:
I can see this mirrors some of the same validation we do prior to markup generation, can maybe get rid of it?